### PR TITLE
[CI]: Refactor: Unify Qwen3-Omni and S2-Pro CI

### LINKED ISSUE #286: [Good First Issue] Refactor: Unify Qwen3-Omni and S2-Pro CI into a shared, parameterized composite action
<h3>Background</h3>
<p>In two recent PRs we staged the benchmark CI for S2-Pro and Qwen3-Omni separately:</p>
<ul>
<li>sgl-project/sglang-omni#278 — [CI] Split S2-Pro Benchmark CI into Multiple Stage Jobs</li>
<li>sgl-project/sglang-omni#283 — [CI] Separated qwen3-omni CI stage</li>
</ul>
<p>Each PR introduced its own composite action (<code>s2pro-setup</code>, <code>qwen3-omni-setup</code>) and its own workflow file. The two setup actions are <strong>~90% identical</strong> (venv restore / dependency install / flashinfer cleanup / GPU process cleanup), and the workflow-level structure is nearly the same. As more models (Ming-Omni, etc.) adopt the same staged CI pattern, this copy-paste will keep accumulating and maintenance cost grows linearly.</p>
<hr>
<h3>Current Divergence</h3>
<p>The two composite actions actually differ in only three places:</p>

Difference | s2pro-setup | qwen3-omni-setup
-- | -- | --
Venv directory | omni-s2pro | omni-qwen3
pip extras | .[s2pro,dev] | .[qwen3,dev] (or equivalent)
torch check in Validate | ❌ Missing | ✅ Present

<p>The last item is a real drift — the <code>import torch</code> sanity check landed in Qwen3-Omni via review feedback on PR #283, but the equivalent step in S2-Pro was not kept in sync. This is the classic downside of having two copies evolve independently.</p>
<hr>
<h3>Goal</h3>
<p>Create a single parameterized composite action that serves all staged benchmark CIs (S2-Pro / Qwen3-Omni / future models), eliminating duplication.</p>
<hr>
<h3>Proposed Approach</h3>
<ol>
<li>Create <code>.github/actions/omni-setup/action.yaml</code> that accepts:
<ul>
<li><code>venv-name</code> (e.g. <code>omni-s2pro</code>, <code>omni-qwen3</code>)</li>
<li><code>pip-extras</code> (e.g. <code>s2pro,dev</code>, <code>qwen3,dev</code>)</li>
<li><code>install-deps</code> (keep the existing boolean)</li>
</ul>
</li>
<li>Move all steps (restore cache, install deps, validate, flashinfer cleanup, GPU cleanup) into this generic action, templating venv paths and pip extras from inputs.</li>
<li>Make the Validate step <strong>always</strong> run the <code>import torch</code> check (adopted from qwen3-omni-setup — default to the stricter variant).</li>
<li>Update both workflow files to <code>uses:</code> the new generic action with matching inputs.</li>
<li>Remove <code>.github/actions/s2pro-setup/</code> and <code>.github/actions/qwen3-omni-setup/</code>.</li>
<li>Run a full CI on the staging runner and confirm both S2-Pro and Qwen3-Omni pipelines are green.</li>
</ol>
<hr>
<h3>Stretch Goals (optional)</h3>
<ul>
<li>Further abstract the workflow-level stage chain (docs → stage-1 → stage-2 → stage-3) as a reusable workflow (<code>workflow_call</code>) rather than duplicated job lists.</li>
<li>Centralize the "keep in sync" env var names (<code>SGLANG_SEEDTTS_MINI_DIR</code>, <code>DATASET_CACHE_ENV</code>) currently duplicated between test files and workflow YAML, so they come from a single source instead of being kept aligned manually.</li>
</ul>
<hr>
<h3>Acceptance Criteria</h3>
<ul>
<li>[ ] A single <code>.github/actions/omni-setup/</code> composite action replaces both existing ones.</li>
<li>[ ] <code>test-s2pro-ci.yaml</code> and <code>test-qwen3-omni-ci.yaml</code> both consume it with their respective inputs.</li>
<li>[ ] Validate step includes the <code>import torch</code> sanity check for both models.</li>
<li>[ ] Both CI pipelines pass end-to-end on the self-hosted runner.</li>
<li>[ ] Net diff removes more lines than it adds (duplication eliminated).</li>
</ul>
<hr>
<h3>Getting Started</h3>
<ul>
<li>Start by reading the full diffs of PR #278 and PR #283 to understand both pipeline shapes.</li>
<li>Changes should be confined to <code>.github/</code> plus workflow <code>uses:</code> references — no Python test changes needed.</li>
<li>CI runs on a self-hosted GPU runner; validate on a fork's staging workflow before requesting review.</li>
<li>Questions welcome in this issue — ping @zhaochenyang20 / @Ratish1 for review.</li>
</ul>
<hr>
<h3>Related</h3>
<ul>

#278 
#283 
#287

</ul></body></html>

### PR DESCRIPTION
**Pull request by nagisa-kunhah**
## Motivation

Issue #286

## Modifications

- Add a shared composite action at `.github/actions/omni-setup/action.yaml`
- Parameterize setup with `venv-name`, `pip-extras`, and `install-deps`
- Update `test-s2pro-ci.yaml` to use the shared action for all staged jobs
- Update `test-qwen3-omni-ci.yaml` to use the shared action instead of inline setup steps
- Unify validation so cached environments always run the `import torch` sanity check
- Remove the old `.github/actions/s2pro-setup/action.yaml`
- Keep Qwen3 cache save behavior consistent by replacing the old cache directory before copying
- Slightly reorder the Qwen3 setup flow to match the shared action: `Restore cache -> Install dependencies/validate -> Remove flashinfer -> Kill GPU processes`

## Related Issues

Closes #286

## Accuracy Test

N/A. This PR only changes GitHub Actions workflow/configuration and does not modify model-side runtime or kernels.

## Benchmark & Profiling

N/A. This PR is a CI refactor and does not change model execution logic or benchmark thresholds.

## Checklist

- [ ] Format your code according with pre-commit.
- [ ] Add unit tests.
- [ ] Update documentation / docstrings / example tutorials as needed.
- [ ] Provide throughput / latency benchmark results and accuracy evaluation results as needed.
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.

### CODE REVIEW

**Comment by zhaochenyang20** [comment]
you guys works so quick

**Comment by nagisa-kunhah** [comment]
> you guys works so quick

Thanks! We probably still need to check the CI results and debug a bit based on them, since we don’t have a downstream runner at the moment.

File: .github/actions/omni-setup/action.yaml [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+inputs:
+  venv-name:
+    description: Name of the virtualenv to restore and activate.
+    required: true
+  pip-extras:
+    description: Optional dependency extras to install, for example `dev` or `s2pro,dev`.
+    required: false
+    default: ""
+  install-deps:
+    description: Prepare or refresh the requested virtualenv before running tests.
```
Comment (zhaochenyang20) [inline .github/actions/omni-setup/action.yaml:13]: `pip-extras` defaults to empty string `""`. If a caller forgets to pass this input (e.g., in a newly added CI job), the Install dependencies step silently falls back to `uv pip install -v -e .` (no extras), which for qwen3 would mean no `pytest`, `jiwer`, or other test dependencies. This error won't surface until pytest runs fail with `ModuleNotFoundError`, making it hard to diagnose.

Suggestion: When `install-deps == 'true'` and `pip-extras` is empty, emit a warning message rather than silently installing the bare package.

**Fix:**
```yaml
      run: |
        ...
        source "${VENV_NAME}/bin/activate"
        if [ -n "${PIP_EXTRAS}" ]; then
          uv pip install -v -e ".[${PIP_EXTRAS}]"
        else
          echo "::warning::pip-extras is empty; installing bare package without extras"
          uv pip install -v -e .
        fi
```

File: .github/actions/omni-setup/action.yaml
Comment (zhaochenyang20) [inline .github/actions/omni-setup/action.yaml:1]: The original qwen3 inline steps ordered: Restore cache → **Remove flashinfer** → Install deps → Kill GPU. The new unified action uses: Restore cache → Install deps → **Remove flashinfer** → Kill GPU (inherited from the original s2pro-setup order).

This is almost certainly immaterial (flashinfer cache removal affects runtime, not install-time), but it is a behavioral change that should be explicitly acknowledged in the PR description for future traceability.

**Review by zhaochenyang20 — CHANGES_REQUESTED** [review]

**Comment by zhaochenyang20** [comment]
@nagisa-kunhah please rebase with main. thanks!

File: .github/actions/omni-setup/action.yaml (hunk 1 above)
Comment (nagisa-kunhah) [reply .github/actions/omni-setup/action.yaml:13]: Fixed, though I’m not sure yet whether it’s worth splitting this into a separate validation step to validate all the inputs.

File: .github/actions/omni-setup/action.yaml
Comment (nagisa-kunhah) [reply .github/actions/omni-setup/action.yaml:1]: OK, I’ve added a note about the Qwen3 step reordering in the PR description.

**Comment by zhaochenyang20** [comment]
strongly think we should first merge https://github.com/sgl-project/sglang-omni/pull/287 then #288

**Comment by nagisa-kunhah** [comment]
> strongly think we should first merge #287 then #288

Got it, no worries.

**Comment by nagisa-kunhah** [comment]
> strongly think we should first merge #287 then #288

@zhaochenyang20 One question: should #288 also absorb the changes from #287 as part of the unify work？

**Comment by zhaochenyang20** [comment]
> > strongly think we should first merge #287 then #288
> 
> @zhaochenyang20 One question: should #288 also absorb the changes from #287 as part of the unify work？

I feel sad to say, that you may need to redo your work after #287, since #287 has a large change 😂

File: .github/actions/omni-setup/action.yaml [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+      if: ${{ inputs.install-deps != 'true' }}
+      shell: bash
+      env:
+        VENV_NAME: ${{ inputs.venv-name }}
+      run: |
+        test -x "${VENV_NAME}/bin/python" || {
+          echo "Prepared ${VENV_NAME} environment is missing. Run the setup stage first."
+          exit 1
+        }
+        "${VENV_NAME}/bin/python" -c "import torch"
```
Comment (zhaochenyang20) [inline .github/actions/omni-setup/action.yaml]: in #287 

`qwen3-omni-setup`'s `Validate prepared environment` step checks both `test -x omni-qwen3/bin/python` AND runs `import torch` validation. This is stricter than `s2pro-setup` (which only checks `test -x`), and is a beneficial improvement. The `Install dependencies` step similarly adds a torch import check to detect corrupted caches. Both improvements are worth adopting in the shared action during #288 unification.

File: .github/workflows/test-qwen3-omni-ci.yaml [hunk 3]
```diff
@@ -150,4 +121,5 @@ jobs:
 
       - name: Save cache
         run: |
+          rm -rf /github/home/omni-qwen3
           cp -p -r omni-qwen3 /github/home/
```
Comment (zhaochenyang20) [inline .github/workflows/test-qwen3-omni-ci.yaml]: Qwen3's "Save cache" step lacks the `if: always()` condition. This means if any earlier test step fails, the cache won't be saved, forcing a full reinstall on the next run. In contrast, S2-Pro's docs job "Save cache" step uses `if: always()` (line 49). This is a **pre-existing issue** (also absent on main), but since this PR's goal is to unify CI behavior across both pipelines, this was an opportunity to fix it. Note: if a test corrupts the venv mid-run, saving a corrupted cache is safe because the Install step detects corruption via `import torch` failure and rebuilds. So adding `if: always()` has no downside.

**Fix:**
```yaml
      - name: Save cache
        if: always()
        run: |
          rm -rf /github/home/omni-qwen3
          cp -p -r omni-qwen3 /github/home/
```

Comment (zhaochenyang20) [reply .github/workflows/test-qwen3-omni-ci.yaml]: also, The "Save cache" logic (`rm -rf /github/home/${VENV_NAME} && cp -p -r ${VENV_NAME} /github/home/`) is still manually inlined in both workflows. While the codebase context explains why it wasn't folded into the composite action (it runs at job end, needs job-specific `if: always()` policy), this rationale is insufficient — the composite action could accept a `save-cache` input, or a second `omni-save-cache` action could be created. The fact that S2-Pro docs has `if: always()` while Qwen3 doesn't is precisely because these steps are scattered and lack centralized management.

However, given the PR description mentions stretch goals including "centralizing env var names" and "workflow_call reusable workflow", and the current scope already achieves the primary objective, this can be tracked as a follow-up.

File: .github/actions/omni-setup/action.yaml [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+          rm -rf "${VENV_NAME}"
+        fi
+        if [ ! -d "${VENV_NAME}" ]; then
+          uv venv "${VENV_NAME}" -p 3.11
+        fi
+        source "${VENV_NAME}/bin/activate"
+        if [ -n "${PIP_EXTRAS}" ]; then
+          uv pip install -v -e ".[${PIP_EXTRAS}]"
+        else
+          echo "::warning::pip-extras is empty; installing bare package without extras"
```
Comment (zhaochenyang20) [inline .github/actions/omni-setup/action.yaml:47]: The `::warning::pip-extras is empty; installing bare package without extras` warning only fires when `install-deps == 'true'` AND `pip-extras` is empty. In current callers, all `install-deps: "true"` invocations explicitly pass `pip-extras` (Qwen3 passes `dev`, S2-Pro passes `s2pro,dev`). Validate-only paths (stages 1/2/3) have `install-deps` defaulting to `"false"`, so the entire Install step is skipped. Thus this warning never fires in any current usage — it's purely defensive against future misuse.

**Review by zhaochenyang20 — CHANGES_REQUESTED** [review]

**Review by zhaochenyang20 — APPROVED** [review]

**Comment by zhaochenyang20** [comment]
@nagisa-kunhah thanks so much for your contribution. You can join SGLang slack with slack.sglang.ai and go to #sglang-omni-dev for discussion.
